### PR TITLE
Serialize the endpoint join/leave at the sandbox level

### PR DESCRIPTION
**Pull request by sanimej**
Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

File: sandbox.go [hunk 1]
```diff
@@ -394,26 +395,29 @@ func (sb *sandbox) clearNetworkResources(ep *endpoint) error {
...
 	)
-	for i, e = range sb.endpoints {
+	for i, e := range sb.endpoints {
 		if e == ep {
-			break
+			index = i
+		}
+		if len(e.Gateway()) > 0 && gwepBefore == nil {
+			gwepBefore = e
 		}
```
Comment (mrjana) [inline sandbox.go]: Shouldn't you break from the loop once you have found the `ep` to delete and `gwepBefore`? There is no reason to traverse the whole list.

File: sandbox.go [hunk 2]
```diff
@@ -394,26 +395,29 @@ func (sb *sandbox) clearNetworkResources(ep *endpoint) error {
...
 	}
-	heap.Remove(&sb.endpoints, i)
-	var highEpAfter *endpoint
-	if len(sb.endpoints) > 0 {
-		highEpAfter = sb.endpoints[0]
+	heap.Remove(&sb.endpoints, index)
+	for _, e := range sb.endpoints {
+		if len(e.Gateway()) > 0 {
+			gwepAfter = e
+		}
```
Comment (mrjana) [inline sandbox.go]: Same here. You should bail out from the loop once you have `gwepAfter`

File: sandbox.go (hunk 1 above)
Comment (sanimej) [inline sandbox.go]: @mrjana This loop also gets the index of the endpoint to be removed. Thats why we have to go through the list fully.

File: sandbox.go (hunk 2 above)
Comment (sanimej) [inline sandbox.go]: Yes, I think here we don't have to go through the full list. will change it.

**Comment by mrjana** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM

**Pull request merged**